src/helpers/ControlledPageComponent.ts

```diff
@@ -0,0 +1,219 @@
+   * be aware that user entered data may be overwritten when the
+   * {@link Database} operation does settle.
+   */
+  disabled: boolean;
+}
+
+/**
+ * A React component that handles {@link ControlledProps}.
+ *
+ * @typeparam Props - The proptypes excluding {@link ControlledProps}.
+ *
+ * @typeparam Value - The type of the value this component links to in the
+ * {@link Database}.
+ */
+export type ControlledComponentType<Props, Value> = React.ComponentType<
+  Props & ControlledProps<Value>
+>;
+
+/**
+ * The options for {@link ControlledPageComponent}.
+ */
+export interface ControlledPageComponentOptions<
+  ComponentType extends ControlledComponentType<
+    Props,
+    StoreValue<DBSchema["schema"], StoreName>
+  >,
+  Props,
+  DBSchema extends NamedSchema<string, number, Schema>,
+  StoreName extends StoreNames<DBSchema["schema"]>
+> {
+  /**
+   * A unique identifier for this component on the page.
+   */
+  key: React.Key;
+
+  /**
+   * The component class or function to render. The component must implement
+   * {@link ControlledProps} in addition to any of its own props (passed through
+   * via {@link ControlledPageComponentOptions.props}).
+   */
+  Component: ComponentType;
+
+  /**
+   * The props to pass to {@link ControlledPageComponentOptions.Component},
+   * excluding {@link ControlledProps}.
+   */
+  props: Props;
+
+  /**
+   * The configuration telling the {@link ControlledPageComponent} how to interface
+   * with the {@link Database}.
+   *
+   * See {@link ControlledPageComponentDatabaseMap} for more details.
+   */
+  databaseMap: ControlledPageComponentDatabaseMap<DBSchema, StoreName>;
+
+  /**
+   * The value to default to when nothing exists in the provided database store.
+   *
+   * If you leave this unspecified, it will use whatever the default value for
+   * the component is.
+   */
+  defaultValue?: StoreValue<DBSchema["schema"], StoreName> | null;
+}
+
+/**
+ * A component along with the associated properties needed to render it as part
+ * of a {@link Step}.
+ *
+ * Unlike for {@link PageComponent}, we inject some additional
+ * {@link ControlledProps} into the component to allow interfacing with a
+ * {@link Database}. The {@link ControlledPageComponentOptions.Component} needs to
+ * support those extra props if they need to interface with a {@link Database}.
+ * If you don't need to this component to interface with a {@link Database},
+ * use a {@link PageComponent} instead.
+ *
+ * ```ts
+ * const storeName = "myStore";
+ *
+ * const myInput = new ControlledPageComponent({
+ *   key: "my-input",
+ *   Component: MyInput,
+ *   props: {
+ *     className: "my-input-class"
+ *   },
+ *   defaultValue: "Nothing here...",
+ *   databaseMap: new ControlledPageComponentDatabseMap<
+ *     MyDBSchema,
+ *     typeof storeName
+ *   >({
+ *     storeName,
+ *     key: "input-0"
+ *   })
+ * });
+ * ```
+ */
+export class ControlledPageComponent<
```

Comment:
This might be clearer if it calls out that it's use case.

Maybe `StaticPageComponent` vs `DynamicPageComponent`?